TraderJoe-Bots/Re-entrancy-Transfers/src/agent.spec.ts

```diff
@@ -0,0 +1,392 @@
+import { provideHandleTransaction } from "./agent";
+import MarketsFetcher from "./markets.fetcher";
+import NetworkData from "./network";
+import { FUNCTIONS_ABIS, MARKET_UPDATE_ABIS } from "./utils";
+
+describe("Re-entrancy transfers bot test suite", () => {
+  let handleTransaction: HandleTransaction;
+
+  const USER_ADDR = createAddress("0xc1");
+  const TEST_MARKETS = [createAddress("0x11"), createAddress("0x22"), createAddress("0x33")];
+  const FUNCTIONS_IFACE = new Interface(FUNCTIONS_ABIS);
+  const mockNetworkManager: NetworkData = {
+    joeTroller: createAddress("0x0ae"),
+    sJoeStaking: createAddress("0x1ae"),
+    masterChefV2: createAddress("0x2ae"),
+    moneyMaker: createAddress("0x3ae"),
+    networkMap: {},
+    setNetwork: jest.fn(),
+  };
+  const mockProvider = new MockEthersProvider();
+  const mockFetcher = new MarketsFetcher(mockProvider as any);
+  const mockUpdateMarkets = jest.fn();
+
+  const createTrace = (from: string, traceAddress: number[], to: string, input: string): Trace => {
+    return {
+      action: {
+        from,
+        to,
+        input,
+      } as TraceAction,
+      traceAddress,
+    } as Trace;
+  };
+
+  const createFinding = (
+    from: string,
+    address: string,
+    rentrancyFrom: string,
+    initialCall: string,
+    reEtrantCall: string
+  ) => {
+    return Finding.fromObject({
+      name: "Re-entrancy detected on a Trader Joe contract",
+      description: "A function call to a trader Joe contract resulted in another call to the same contract",
+      alertId: "TraderJoe-25",
+      protocol: "TraderJoe",
+      severity: FindingSeverity.High,
+      type: FindingType.Exploit,
+      metadata: {
+        from: from, // address initializing the call.
+        initialCall: initialCall, // functions that was called in our contract and resulted in a re-entrancy.
+        entrancyFrom: rentrancyFrom, // the contract re-entring our contract.
+        reEtrantCall: reEtrantCall, // second call to the same traderJoe contract.
+      },
+      addresses: [address],
+    });
+  };
+
+  beforeAll(() => {
+    handleTransaction = provideHandleTransaction(mockNetworkManager as any, mockFetcher);
+    mockFetcher.markets = new Set(TEST_MARKETS);
+    mockFetcher.updateMarkets = mockUpdateMarkets;
+  });
+
+  beforeEach(() => mockUpdateMarkets.mockClear());
+
+  it("should ignore empty transactions", async () => {
+    const txEvent: TransactionEvent = new TestTransactionEvent();
+
+    const findings = await handleTransaction(txEvent);
+    expect(findings).toStrictEqual([]);
+  });
+
+  it("should ignore calls when no re-entrancy happens", async () => {
+    const traces = [
+      createTrace(
+        USER_ADDR,
+        [],
+        mockNetworkManager.sJoeStaking,
+        FUNCTIONS_IFACE.encodeFunctionData("withdraw(uint256)", [BigNumber.from("1000")])
+      ),
+    ];
+
+    const txEvent: TransactionEvent = createTransactionEvent({
```

Comment:
Any particular reason to use `createTransactionEvent` over `TestTransactionEvent.addTrace()` from the `GAM`? Just curious.

Reply:
That's because the `addTrace` from `GAM` doesn't support the `traceAddress` field, which is used in our logic.